src/main/kotlin/io/github/mojira/arisa/JiraConnectionService.kt

```diff
@@ -0,0 +1,134 @@
+package io.github.mojira.arisa
+
+import com.uchuhimo.konf.Config
+import io.github.mojira.arisa.infrastructure.config.Arisa
+import io.github.mojira.arisa.infrastructure.jira.connectToJira
+import java.lang.Long.max
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+
+class JiraConnectionService(
+    private val config: Config
+) {
+    companion object {
+        /**
+         * How long ago the last successful connection needs to be for Arisa to try to relog
+         */
+        private const val MAX_SECONDS_SINCE_LAST_SUCCESSFUL_CONNECTION = 60L
+
+        /**
+         * How long Arisa should wait after it was unable to relog
+         */
+        private const val WAIT_TIME_AFTER_UNSUCCESSFUL_RELOG_IN_SECONDS = 5 * 60L // 5 minutes
+
+        /**
+         * How long Arisa should wait after a connection error before continuing without trying to relog
+         */
+        private const val WAIT_TIME_AFTER_CONNECTION_ERROR_IN_SECONDS = 40L
+    }
+
+    private var lastSuccessfulConnection = Instant.now().minusSeconds(MAX_SECONDS_SINCE_LAST_SUCCESSFUL_CONNECTION + 1)
+
+    /**
+     * Tries to establish a connection and log into Jira.
+     * @returns null if able to connect, the connection exception otherwise
+     */
+    @Suppress("TooGenericExceptionCaught")
+    private fun establishConnection(): Exception? {
+        return try {
+            val client = connectToJira(
+                config[Arisa.Credentials.username],
+                config[Arisa.Credentials.password],
+                config[Arisa.Issues.url]
+            )
+
+            log.info("Successfully connected to jira")
+            lastSuccessfulConnection = Instant.now()
+            jiraClient = client
+
+            null
+        } catch (exception: Exception) {
+            exception
+        }
+    }
+
+    /**
+     * Try relogging if the last successful connection was at least [MAX_SECONDS_SINCE_LAST_SUCCESSFUL_CONNECTION] ago.
+     * @return
+     * - [RelogResult.SuccessfulRelog] if relog was successful
+     * - [RelogResult.UnsucessfulRelog] if unable to relog
+     * - [RelogResult.NoRelogAttempted] if no relog was attempted because the last relog happened recently
+     */
+    fun tryRelog(): RelogResult {
+        val secondsSinceLastSuccessfulConnection = Duration.between(lastSuccessfulConnection, Instant.now()).toSeconds()
+
+        return if (secondsSinceLastSuccessfulConnection > MAX_SECONDS_SINCE_LAST_SUCCESSFUL_CONNECTION) {
+            log.info("Trying to relog")
+
+            establishConnection() ?: return RelogResult.SuccessfulRelog()
```

Comment:
For sanity should probably call `notifyOfSuccessfulConnection()` after successful relog. Otherwise if for some other reason execution keeps failing (e.g. broken module) Arisa would continuously relog (with only `MIN_TIME_BETWEEN_EXECUTION_CYCLES_IN_SECONDS` delay in between).

Reply:
Yeah, I wish we wouldn't need `notifyOfSuccessfulConnection`, it seems like it's risky that we might forget it somewhere. And of course I did.